### PR TITLE
Move Steal to rustc_data_structures.

### PR DESCRIPTION
**Pull request by cjgillot**

### CODE REVIEW

**Comment by jyn514** [comment]
@cjgillot this is meant to improve compile times for rustc_middle?

**Comment by cjgillot** [comment]
For example. 
Second motivation : `Steal` is just a data structure, with no dependency on rustc_middle.
`rustc_data_structures` just feels like a better place.

File: compiler/rustc_data_structures/src/steal.rs [hunk 1]
```diff
@@ -31,7 +32,7 @@ impl<T> Steal<T> {
 
     pub fn borrow(&self) -> MappedReadGuard<'_, T> {
         ReadGuard::map(self.value.borrow(), |opt| match *opt {
-            None => bug!("attempted to read from stolen value"),
+            None => panic!("attempted to read from stolen value"),
```
Comment (pickfire) [inline compiler/rustc_data_structures/src/steal.rs:35]: Is this change related?

Comment (jyn514) [reply compiler/rustc_data_structures/src/steal.rs:35]: Yes, `bug` is only defined in rustc_middle: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/macro.bug.html
